### PR TITLE
Task/fill egresos ingresos form

### PR DESCRIPTION
**Pull request by AndreeToledo**
#### What's this PR do?
Allows a capturista to add and edit, ingresos and egresos for a family

#### Where should the reviewer start?
First check the captura/views.py
Then check the forms

#### How should this be manually tested?
Login as capturista
Then go to edit a studio
Go to the transacciones page
Play with ingresos y egresos

#### Any background context you want to provide?

This template was adapted ~stolen~ from [Quickleft/Sprint.ly](https://quickleft.com/blog/pull-request-templates-make-code-review-easier/)

### CODE REVIEW

File: jp2_online/static/js/captura/captura_transaccion.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+          confirmButtonText: 'OK'
+        });
+      }
+    });
+  });
+}
+
+
+$(document).ready(function() {
+  // hide inputs from creation form
```
Comment (csssaz) [inline jp2_online/static/js/captura/captura_transaccion.js]: remove comment

File: jp2_online/static/js/captura/captura_transaccion.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
+console.log('saludos')
```
Comment (csssaz) [inline jp2_online/static/js/captura/captura_transaccion.js]: ?

File: jp2_online/static/js/captura/captura_transaccion.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+function changeSubmitToAjax(id_modal, id_form) {
+  console.log(id_form)
```
Comment (csssaz) [inline jp2_online/static/js/captura/captura_transaccion.js]: ?

File: indicadores/forms.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+class IngresoForm(ModelForm):
+    """ Model form for Ingreso
+
+    This is the general form for creating and updating an
+    ingreso.
+    """
+
+    class Meta:
+        model = Ingreso
+        fields = ('fecha',
```
Comment (csssaz) [inline indicadores/forms.py]: this should have some sort of explanation in the label. It's not obvious what is it referring to

File: familias/utils.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import decimal
+from django.shortcuts import get_object_or_404
+from indicadores.models import Transaccion
+from .models import Familia
+
+
+def total_egresos_familia(id_familia):
```
Comment (csssaz) [inline familias/utils.py]: all of these should be tested.

File: familias/utils.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    transacciones = Transaccion.objects.filter(familia=familia,
+                                               activo=True,
+                                               es_ingreso=False)
+    total = decimal.Decimal('0')
+    for transaccion in transacciones:
+        total = total + transaccion.obtener_valor_mensual()
+
+    return total;
+
+def total_ingresos_familia(id_familia):
```
Comment (csssaz) [inline familias/utils.py]: Also it isn't necessary but I would just create one function which receives the es_ingreso, and depending on its value make the query for the transactions (if it is None, you don't pass it, for total_neto). And the total can be a one liner using sum().

File: jp2_online/static/js/captura/captura_transaccion.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+var csrftoken = getCookie('csrftoken');
+$.ajaxSetup({
+  beforeSend: function(xhr, settings) {
+    if (!csrfSafeMethod(settings.type) && !this.crossDomain) {
+      xhr.setRequestHeader('X-CSRFToken', csrftoken);
+    }
+  }
+});
+
+function initEditForm() {
```
Comment (csssaz) [inline jp2_online/static/js/captura/captura_transaccion.js]: mini comment would be useful

File: captura/views.py [hunk 8]
```diff
@@ -410,6 +412,95 @@ def get_form_edit_integrante(request, id_integrante):
...
+
+@login_required
+@user_passes_test(is_capturista)
+def delete_transaccion(request):
+    pass
+
+
+@login_required
+@user_passes_test(is_capturista)
+def transacciones(request, id_familia):
```
Comment (csssaz) [inline captura/views.py]: please change the name to something like 'list_transacciones' which gives more info and is more readable.

File: captura/views.py [hunk 9]
```diff
@@ -410,6 +412,95 @@ def get_form_edit_integrante(request, id_integrante):
...
+                                                      initial={'id_transaccion': id_transaccion})
+        if hasattr(transaccion, 'ingreso'):
+            context['ingreso_form'] = IngresoForm(id_familia, instance=transaccion.ingreso)
+        return render(request, 'captura/edit_ingreso_egreso_form.html', context)
+    return HttpResponseBadRequest()
+
+
+@login_required
+@user_passes_test(is_capturista)
+def delete_transaccion(request):
```
Comment (csssaz) [inline captura/views.py]: also a simple docstring telling that "this view is meant to bla bla"

File: captura/views.py [hunk 10]
```diff
@@ -410,6 +412,95 @@ def get_form_edit_integrante(request, id_integrante):
...
+        response_data = {}
+        if request.POST['id_transaccion']:  # In case of updating
+            transaccion = get_object_or_404(Transaccion, pk=request.POST['id_transaccion'])
+            transaccion_form = TransaccionForm(request.POST, instance=transaccion)
+        else:  # In case of creation
+            transaccion_form = TransaccionForm(request.POST)
+        if transaccion_form.is_valid():
+            transaccion_form.save()
+            if transaccion_form.cleaned_data['es_ingreso']:
+                if hasattr(transaccion_form.instance, 'ingreso'):
```
Comment (csssaz) [inline captura/views.py]: please add a short comment indicating that this is in case of updating (as you did above)

File: captura/views.py [hunk 11]
```diff
@@ -410,6 +412,95 @@ def get_form_edit_integrante(request, id_integrante):
...
+                    ingreso_form = IngresoForm(id_familia, request.POST,
+                                               instance=transaccion_form.instance.ingreso)
+                else:
+                    ingreso_form = IngresoForm(id_familia, request.POST)
+                if ingreso_form.is_valid():
+                    ingreso = ingreso_form.save(commit=False)
+                    ingreso.transaccion = transaccion_form.instance
+                    print(ingreso_form.errors)
```
Comment (csssaz) [inline captura/views.py]: ?

File: indicadores/forms.py [hunk 12]
```diff
@@ -43,6 +43,9 @@ class Meta:
         fields = ('fecha',
                   'tipo',
                   'tutor')
+        labels = {
+            'fecha': 'Desde cuando recive el ingreso?'
```
Comment (csssaz) [inline indicadores/forms.py]: 👎 👎 👎 👎 👎 👎 👎 👎 👎 🆘🆘🆘🆘🆘🆘🆘🆘🆘🆘🆘🆘

Comment (csssaz) [inline indicadores/forms.py]: ¿Desde cuándo recibe el ingreso?

File: captura/test_views.py [hunk 13]
```diff
@@ -837,49 +843,153 @@ def setUp(self):
...
+
+        r = self.client.post(reverse('captura:create_transaccion',
+                                     kwargs={'id_familia': self.familia1.id}),
+                             data=self.transaccion_constructor_dictionary,
+                             HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        content = json.loads(r.content.decode('utf-8'))
+        self.assertEqual(content['msg'], 'Egreso guardado con éxito')
+        self.assertEqual(r.status_code, 200)
+
+    def test_create_egreso_incomplete(self):
```
Comment (csssaz) [inline captura/test_views.py]: please add simple docstrings to these tests explaining what they do

File: jp2_online/templates/captura/dashboard_transacciones.html [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+          </div>
+        </div>
+      </div>                
+    </div>
+  </div>
+</div>
+
+<div class="text-center">
+  <a href="{% url 'captura:list_integrantes' familia.id %}" class="btn btn-danger">
+    <i class="glyphicon glyphicon-arrow-left"></i> Ingresos y Egresos
```
Comment (fernandolobato) [inline jp2_online/templates/captura/dashboard_transacciones.html]: This should say Integrantes

File: jp2_online/templates/captura/egreso_form.html [hunk 15]
```diff
@@ -0,0 +1,4 @@
+<a class="btn btn-primary" href="{% url 'captura:create_egreso' %}" id="create_estudio">
+    <i class="glyphicon glyphicon-plus"></i>
+    Agregar egreso
+</a>
```
Comment (fernandolobato) [inline jp2_online/templates/captura/egreso_form.html]: Are you sure we need to separate one button into a file by itself?? I think this add unnecessary complexity. The captura templates folders already has too many files . Besides... this is not a form, it is just a button.

Comment (AndreeToledo) [inline jp2_online/templates/captura/egreso_form.html]: They can be merged back, this separate structure was implemented when creating wasn't separated from editing, I will return this to the dashboard_transaccion

Comment (fernandolobato) [inline jp2_online/templates/captura/egreso_form.html]: 👍

**Pull request merged**